auth.py

```diff
@@ -0,0 +1,304 @@
+            return
+        
+        print("BinarySecurityToken received.")
+        self.token = binarySecurityToken
+
+        return True
+    
+    def _get_cookie(self):
+        # regex will get the 'tenant url' from the site provided for login
+        SPO_IDCRL_URL = "https://" + re.search('(.+\.com)/', self.site).group(1) + "/_vti_bin/idcrl.svc/"
+        
+        # HTTP Post request with authorization headers
+        headers = {"Authorization": "BPOSIDCRL " + self.token, 
+                   "X-IDCRL_ACCEPTED":"t", 
+                   "User-Agent": "" }
+        response = requests.get(url=SPO_IDCRL_URL, headers=headers)
+
+        if response.status_code == requests.codes.ok:
+            print("ADFS Authentication successful")
+
+            # Add the SPOIDCRL cookie to the session
+            self.cookie = self._buildcookie(response.cookies)
+            
+            
+            return True
+        else:
+            print("ADFS Authentication failed")
+
+    def _get_digest(self):
+        """Check and refresh sites cookie and request digest"""
+        
+        if self.expire <= datetime.now():
+            # Template for SOAP digest request using <tenant>.sharepoint.com/sites/<sub site>/_vti_bin/sites
+            headers = ({"SOAPAction": "http://schemas.microsoft.com/sharepoint/soap/GetUpdatedFormDigestInformation",
+                                "Host" : self.tenantBaseURL,
+                                "Content-Type" : "text/xml;charset=utf-8",
+                                "Cookie" : self.cookie,
+                                "X-RequestForceAuthentication" : "true"})
+            digestEnvelope = '<?xml version="1.0" encoding="utf-8"?><soap:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/"><soap:Body><GetUpdatedFormDigest xmlns="http://schemas.microsoft.com/sharepoint/soap/" /></soap:Body></soap:Envelope>'
+            
+            # Request site context info from SharePoint site
+            requestUrl = "https://" + self.tenantBaseURL + "/_vti_bin/sites.asmx"
+            response = requests.post(requestUrl, data=digestEnvelope, headers=headers)
+            
+
+            # Parse digest text and timeout from XML
+            try:
+                
+                self.digest = re.search("<DigestValue>(.+)</DigestValue>", response.text).group(1)
+                timeout = int(re.search("Seconds>(\d+)<", response.text).group(1))
+                
+            except:
+                print("Digest request failed")
+                return
+            # Calculate digest expiry time
+            self.expire = datetime.now() + timedelta(seconds=timeout)
+
+        return True
+    def _buildcookie(self, cookies):
+        """Create session cookie from response cookie dictionary"""
+        
+        return "SPOIDCRL=" + cookies["SPOIDCRL"]
```

Comment:
Please add a newline at the end of all files

Reply:
✔️